src/stan_math_backend/Stan_math_code_gen.ml

```diff
@@ -533,11 +604,19 @@ let escape_name str =
 
 let pp_prog ppf (p : (mtype_loc_ad with_expr, stmt_loc) prog) =
   (* First, do some transformations on the MIR itself before we begin printing it.*)
+  let fix_data_reads stmts =
+    { stmt= Decl {decl_adtype= DataOnly; decl_id= pos; decl_type= Sized SInt}
+    ; smeta= no_span }
+    :: stmts
+    |> List.concat_map ~f:invert_read_fors
+    |> List.concat_map ~f:add_pos_reset
+    |> List.map ~f:use_pos_in_readdata
+  in
   let p =
     { p with
       prog_name= escape_name p.prog_name
-    ; prepare_data= List.map ~f:(xform_readdata []) p.prepare_data
-    ; transform_inits= List.map ~f:(xform_readdata []) p.transform_inits }
+    ; prepare_data= fix_data_reads p.prepare_data
```

Comment:
@seantalts , could you explain a bit more about why these need to be fixed rather than generating the correct code in AST_to_mir in the first place?

Reply:
It's the same thing we talked about in the meeting - the API to a Stan Math model class asks for data in the wrong order, so we have to invert the for loops in code gen in a C++ specific way. There are also a few other items that are even more C++ specific in there (using the `pos` variable, etc).

Reply:
Right! So I guess you're saying this is funny behaviour that is specific to this particular backend? That makes sense.

Reply:
Yep. :/ And that raises the question of like, is the current thing where we translate a sized type into MIR for loops a good abstraction for the MIR at all (because other backends might not want that either). But so far it's workable so I'm leaving these more perfectionist philosophical questions off my plate until we launch (though others could take a crack at them).

Reply:
I think that makes a lot of sense. Once everything is in place, we should probably do a big refactor if we're feeling brave.